### PR TITLE
IDENTITY-4099 Add audit logs to privileged operations - Workflow

### PR DESCRIPTION
**Pull request by cdwijayarathna**

### CODE REVIEW

**Comment by johannnallathamby** [comment]
Remove all the try - catch surrounding the listeners. Thats the beauty of this pattern; the business layer doesn't know to handle exceptions arising from the extensions; they just let it be throw outside the method.

File: components/workflow-mgt/org.wso2.carbon.identity.workflow.impl/src/main/java/org/wso2/carbon/identity/workflow/impl/WorkflowImplServiceImpl.java [hunk 1]
```diff
@@ -170,6 +291,17 @@ public void deleteHumanTask(WorkflowRequest workflowRequest) throws WorkflowImpl
...
+            for (WorkflowImplServiceListener workflowListener : workflowListenerList) {
+                try {
+                    workflowListener.doPostDeleteHumanTask(workflowRequest);
+                } catch (WorkflowImplException e) {
+                    throw new WorkflowImplException(
+                            "Error occurred while calling doPostDeleteHumanTask in WorkflowImplServiceListener ," +
+                                    workflowListener.getClass().getName(), e);
+                }
+            }
+
```
Comment (johannnallathamby) [inline components/workflow-mgt/org.wso2.carbon.identity.workflow.impl/src/main/java/org/wso2/carbon/identity/workflow/impl/WorkflowImplServiceImpl.java]: Listeners must come as the last statement of the method; even after the catch block.

File: components/workflow-mgt/org.wso2.carbon.identity.workflow.impl/src/main/java/org/wso2/carbon/identity/workflow/impl/internal/WorkflowImplServiceComponent.java [hunk 2]
```diff
@@ -201,4 +209,18 @@ protected void setIdentityCoreInitializedEventService(IdentityCoreInitializedEve
          is started */
     }
 
+    protected void setWorkflowImplServiceListener(WorkflowImplServiceListener workflowListener) {
+        if(workflowListener !=null) {
```
Comment (johannnallathamby) [inline components/workflow-mgt/org.wso2.carbon.identity.workflow.impl/src/main/java/org/wso2/carbon/identity/workflow/impl/internal/WorkflowImplServiceComponent.java]: Do we need null check here ? I don't think anyone can register a service with null object

File: components/workflow-mgt/org.wso2.carbon.identity.workflow.impl/src/main/java/org/wso2/carbon/identity/workflow/impl/internal/WorkflowImplServiceComponent.java [hunk 3]
```diff
@@ -201,4 +209,18 @@ protected void setIdentityCoreInitializedEventService(IdentityCoreInitializedEve
...
+            WorkflowImplServiceDataHolder.getInstance().getWorkflowListenerList()
+                    .add(workflowListener);
+        }
+    }
+
+    protected void unsetWorkflowImplServiceListener(WorkflowImplServiceListener workflowListener) {
+        if(workflowListener !=null) {
```
Comment (johannnallathamby) [inline components/workflow-mgt/org.wso2.carbon.identity.workflow.impl/src/main/java/org/wso2/carbon/identity/workflow/impl/internal/WorkflowImplServiceComponent.java]: Do we need null check here ? I don't think anyone can register a service with null object. Also formatting is wrong:
!= null

File: components/workflow-mgt/org.wso2.carbon.identity.workflow.impl/src/main/java/org/wso2/carbon/identity/workflow/impl/internal/WorkflowImplServiceDataHolder.java [hunk 4]
```diff
@@ -108,5 +114,14 @@ public WorkflowImplService getWorkflowImplService() {
     public void setWorkflowImplService(WorkflowImplService workflowImplService) {
         this.workflowImplService = workflowImplService;
     }
+
+    public List<WorkflowImplServiceListener> getWorkflowListenerList() {
+        return workflowListenerList;
+    }
+
+    public void setWorkflowListenerList(
```
Comment (johannnallathamby) [inline components/workflow-mgt/org.wso2.carbon.identity.workflow.impl/src/main/java/org/wso2/carbon/identity/workflow/impl/internal/WorkflowImplServiceDataHolder.java]: Shall have this as addWorkflowListener ?

File: components/workflow-mgt/org.wso2.carbon.identity.workflow.impl/src/main/java/org/wso2/carbon/identity/workflow/impl/listener/AbstractWorkflowImplServiceListener.java [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+
+    /**
+     * Trigger after retrieving a BPS profile
+     *
+     * @param profileName
+     * @param tenantId
+     * @throws WorkflowImplException
+     */
+    @Override
+    public void doPostGetBPSProfile(String profileName, int tenantId) throws WorkflowImplException {
```
Comment (johannnallathamby) [inline components/workflow-mgt/org.wso2.carbon.identity.workflow.impl/src/main/java/org/wso2/carbon/identity/workflow/impl/listener/AbstractWorkflowImplServiceListener.java]: In this one and all the workflow listeners, for any get methods' post listeners, can you pass the result of the do method also as the first final argument.

**Comment by johannnallathamby** [comment]
In this one and all the workflow listeners, for any get methods' post listeners, can you pass the result of the do method also as the first final argument.

File: components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/WorkflowManagementServiceImpl.java [hunk 6]
```diff
@@ -149,6 +205,10 @@ public WorkflowEvent getEvent(String id) {
                 }
                 event.setParameters(parameters);
             }
+
+            for (WorkflowListener workflowListener : workflowListenerList) {
+                workflowListener.doPostGetEvent(id);
+            }
```
Comment (johannnallathamby) [inline components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/WorkflowManagementServiceImpl.java]: Take the listener execution outside the condition block. You may have to assign the result to a local variable and return later.

File: components/workflow-mgt/org.wso2.carbon.identity.workflow.impl/src/main/java/org/wso2/carbon/identity/workflow/impl/WorkflowImplServiceImpl.java [hunk 7]
```diff
@@ -174,6 +229,10 @@ public void deleteHumanTask(WorkflowRequest workflowRequest) throws WorkflowImpl
                 RemoteException | IllegalStateFault | IllegalArgumentFault e) {
             throw new WorkflowImplException("Error while deleting the human tasks of the request.");
```
Comment (pulasthi7) [inline components/workflow-mgt/org.wso2.carbon.identity.workflow.impl/src/main/java/org/wso2/carbon/identity/workflow/impl/WorkflowImplServiceImpl.java]: wrap or log 'e'

File: components/workflow-mgt/org.wso2.carbon.identity.workflow.impl/src/main/java/org/wso2/carbon/identity/workflow/impl/internal/WorkflowImplServiceDataHolder.java [hunk 8]
```diff
@@ -23,10 +23,15 @@
...
 import org.wso2.carbon.identity.workflow.impl.WorkflowImplService;
+import org.wso2.carbon.identity.workflow.impl.listener.WorkflowImplServiceListener;
 import org.wso2.carbon.identity.workflow.mgt.WorkflowManagementService;
+import org.wso2.carbon.identity.workflow.mgt.listener.WorkflowListener;
 import org.wso2.carbon.user.core.service.RealmService;
 import org.wso2.carbon.utils.ConfigurationContextService;
 
+import java.util.ArrayList;
+import java.util.List;
+
```
Comment (pulasthi7) [inline components/workflow-mgt/org.wso2.carbon.identity.workflow.impl/src/main/java/org/wso2/carbon/identity/workflow/impl/internal/WorkflowImplServiceDataHolder.java]: remove this unnecessary line

File: components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/WorkflowManagementServiceImpl.java [hunk 9]
```diff
@@ -169,14 +208,25 @@ public WorkflowEvent getEvent(String id) {
                 templates.add(template);
             }
         }
+        for (WorkflowListener workflowListener : workflowListenerList) {
+            workflowListener.doPostListTemplates(templates);
+        }
+
+
```
Comment (pulasthi7) [inline components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/WorkflowManagementServiceImpl.java]: Remove these unnecessary lines

File: components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/WorkflowManagementServiceImpl.java [hunk 10]
```diff
@@ -455,10 +618,20 @@ public boolean areTwoEntitiesRelated(Entity entity1, Entity entity2) throws
...
+    public boolean isEventAssociated(String eventType) throws WorkflowException {
 
-        List<WorkflowAssociation> associations = workflowRequestAssociationDAO.getWorkflowAssociationsForRequest(eventType, CarbonContext
+        List<WorkflowListener> workflowListenerList =
+                WorkflowServiceDataHolder.getInstance().getWorkflowListenerList();
+        for (WorkflowListener workflowListener : workflowListenerList) {
+            workflowListener.doPreIsEventAssociated(eventType);
+        }
+        List<WorkflowAssociation> associations = workflowRequestAssociationDAO.getWorkflowAssociationsForRequest
+                (eventType, CarbonContext
```
Comment (pulasthi7) [inline components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/WorkflowManagementServiceImpl.java]: This line can be merged with the line below

File: components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/listener/WorkflowAuditLogger.java [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+     * @throws WorkflowException
+     */
+    @Override
+    public void doPostDeleteWorkflow(Workflow workflow) throws WorkflowException {
+        String loggedInUser = PrivilegedCarbonContext.getThreadLocalCarbonContext().getUsername();
+        if (StringUtils.isBlank(loggedInUser)) {
+            loggedInUser = CarbonConstants.REGISTRY_SYSTEM_USERNAME;
+        }
+        String auditData = "\"" + "Workflow ID" + "\" : \"" + workflow.getWorkflowId()
+                + "\"";
```
Comment (pulasthi7) [inline components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/listener/WorkflowAuditLogger.java]: Can be appended to the line above

File: components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/listener/WorkflowAuditLogger.java [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+        String loggedInUser = PrivilegedCarbonContext.getThreadLocalCarbonContext().getUsername();
+        if (StringUtils.isBlank(loggedInUser)) {
+            loggedInUser = CarbonConstants.REGISTRY_SYSTEM_USERNAME;
+        }
+        String auditData = "\"" + "Workflow Name" + "\" : \"" + workflowDTO.getWorkflowName()
+                + "\",\"" + "Workflow Impl ID" + "\" : \"" + workflowDTO.getWorkflowImplId()
+                + "\",\"" + "Workerflow ID" + "\" : \"" + workflowDTO.getWorkflowId()
+                + "\",\"" + "Workflow Description" + "\" : \"" + workflowDTO.getWorkflowDescription()
+                + "\",\"" + "Template ID" + "\" : \"" + workflowDTO.getTemplateId()
+                + "\"";
```
Comment (pulasthi7) [inline components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/listener/WorkflowAuditLogger.java]: Can be appended to the line above

File: components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/listener/WorkflowAuditLogger.java [hunk 13]
```diff
@@ -0,0 +1,159 @@
...
+     * @throws WorkflowException
+     */
+    @Override
+    public void doPostRemoveAssociation(int associationId) throws WorkflowException {
+        String loggedInUser = PrivilegedCarbonContext.getThreadLocalCarbonContext().getUsername();
+        if (StringUtils.isBlank(loggedInUser)) {
+            loggedInUser = CarbonConstants.REGISTRY_SYSTEM_USERNAME;
+        }
+        String auditData = "\"" + "Association ID" + "\" : \"" + associationId
+                + "\"";
```
Comment (pulasthi7) [inline components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/listener/WorkflowAuditLogger.java]: Can be appended to the line above

File: components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/listener/WorkflowAuditLogger.java [hunk 14]
```diff
@@ -0,0 +1,159 @@
...
+     */
+    @Override
+    public void doPostChangeAssociationState(String associationId, boolean isEnable) throws WorkflowException {
+        String loggedInUser = PrivilegedCarbonContext.getThreadLocalCarbonContext().getUsername();
+        if (StringUtils.isBlank(loggedInUser)) {
+            loggedInUser = CarbonConstants.REGISTRY_SYSTEM_USERNAME;
+        }
+        String auditData = "\"" + "Association ID" + "\" : \"" + associationId
+                + "\",\"" + "Resulting State" + "\" : \"" + isEnable
+                + "\"";
```
Comment (pulasthi7) [inline components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/listener/WorkflowAuditLogger.java]: Can be appended to the line above

File: components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/listener/WorkflowExecutorAuditLogger.java [hunk 15]
```diff
@@ -0,0 +1,78 @@
...
+     */
+    @Override
+    public void doPostHandleCallback(String uuid, String status, Map<String, Object> additionalParams) throws WorkflowException {
+        String loggedInUser = PrivilegedCarbonContext.getThreadLocalCarbonContext().getUsername();
+        if (StringUtils.isBlank(loggedInUser)) {
+            loggedInUser = CarbonConstants.REGISTRY_SYSTEM_USERNAME;
+        }
+        String auditData = "\"" + "Request ID" + "\" : \"" + uuid
+                + "\",\"" + "Callback Status" + "\" : \"" + status
+                + "\"";
```
Comment (pulasthi7) [inline components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/listener/WorkflowExecutorAuditLogger.java]: Can be appended to the line above

File: components/workflow-mgt/org.wso2.carbon.identity.workflow.impl/src/main/java/org/wso2/carbon/identity/workflow/impl/WorkflowImplServiceImpl.java (hunk 7 above)
Comment (cdwijayarathna) [inline components/workflow-mgt/org.wso2.carbon.identity.workflow.impl/src/main/java/org/wso2/carbon/identity/workflow/impl/WorkflowImplServiceImpl.java]: This has already been fixed with  https://github.com/wso2/carbon-identity/commit/a3d18711e29c26f313b8c8efd9c17302e3434eb1

File: components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/listener/WorkflowExecutorAuditLogger.java (hunk 15 above)
Comment (cdwijayarathna) [inline components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/listener/WorkflowExecutorAuditLogger.java]: Thanks for pointing this, fixed with https://github.com/wso2/carbon-identity/pull/1675

File: components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/listener/WorkflowAuditLogger.java (hunk 14 above)
Comment (cdwijayarathna) [inline components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/listener/WorkflowAuditLogger.java]: Thanks for pointing this, fixed with https://github.com/wso2/carbon-identity/pull/1675

File: components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/listener/WorkflowAuditLogger.java (hunk 13 above)
Comment (cdwijayarathna) [inline components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/listener/WorkflowAuditLogger.java]: Thanks for pointing this, fixed with https://github.com/wso2/carbon-identity/pull/1675

File: components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/listener/WorkflowAuditLogger.java (hunk 12 above)
Comment (cdwijayarathna) [inline components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/listener/WorkflowAuditLogger.java]: Thanks for pointing this, fixed with https://github.com/wso2/carbon-identity/pull/1675

File: components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/listener/WorkflowAuditLogger.java (hunk 11 above)
Comment (cdwijayarathna) [inline components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/listener/WorkflowAuditLogger.java]: Thanks for pointing this, fixed with https://github.com/wso2/carbon-identity/pull/1675

File: components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/WorkflowManagementServiceImpl.java (hunk 10 above)
Comment (cdwijayarathna) [inline components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/WorkflowManagementServiceImpl.java]: Thanks for pointing this, fixed with https://github.com/wso2/carbon-identity/pull/1675

File: components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/WorkflowManagementServiceImpl.java (hunk 9 above)
Comment (cdwijayarathna) [inline components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/WorkflowManagementServiceImpl.java]: Thanks for pointing this, fixed with https://github.com/wso2/carbon-identity/pull/1675

File: components/workflow-mgt/org.wso2.carbon.identity.workflow.impl/src/main/java/org/wso2/carbon/identity/workflow/impl/internal/WorkflowImplServiceDataHolder.java (hunk 8 above)
Comment (cdwijayarathna) [inline components/workflow-mgt/org.wso2.carbon.identity.workflow.impl/src/main/java/org/wso2/carbon/identity/workflow/impl/internal/WorkflowImplServiceDataHolder.java]: Thanks for pointing this, fixed with https://github.com/wso2/carbon-identity/pull/1675

**Pull request merged**